From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.16 october2005, patch 05-03
Patch-Mainline: 2.6.18-rc1

  Description: cio: unable to set DASD root device online
  Symptom:     Trying to set a DASD root device online fails under some
               circumstances with message "Read configuration data returned
               error -5".
  Problem:     Function "read configuration data" incorrectly aborts with -EIO
               when encountering a temporary busy condition at a storage server.
  Solution:    Perform retry when encountering temporary busy conditions.
  Problem-ID:  24232
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/drivers/s390/cio/device_ops.c	18 Jan 2006 11:27:58 -0000	1.62
+++ linux-2.5/drivers/s390/cio/device_ops.c	31 May 2006 12:51:22 -0000	1.62.2.1
@@ -270,7 +270,8 @@
 		 * We didn't get channel end / device end. Check if path
 		 * verification has been started; we can retry after it has
 		 * finished. We also retry unit checks except for command reject
-		 * or intervention required.
+		 * or intervention required. Also check for long busy
+		 * conditions.
 		 */
 		 if (cdev->private->flags.doverify ||
 			 cdev->private->state == DEV_STATE_VERIFY)
@@ -279,6 +280,10 @@
 		     !(irb->ecw[0] &
 		       (SNS0_CMD_REJECT | SNS0_INTERVENTION_REQ)))
 			 cdev->private->intparm = -EAGAIN;
+		else if ((irb->scsw.dstat & DEV_STAT_ATTENTION) &&
+			 (irb->scsw.dstat & DEV_STAT_DEV_END) &&
+			 (irb->scsw.dstat & DEV_STAT_UNIT_EXCEP))
+			cdev->private->intparm = -EAGAIN;
 		 else
 			 cdev->private->intparm = -EIO;
 			 

